Enhance match system and API response handling

haoteng10 opened the pull request:
- Added support for shared constants and standardized API response formats across the backend and frontend.
- Refactored match-related services to utilize new response structures, improving error handling and consistency.
- Updated match polling and telemetry services to use shared configurations for match statuses.
- Implemented token clearing functionality for players disconnecting before match starts, ensuring tokens can be reused.
- Adjusted various components to reflect changes in match token handling and API interactions.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Unifies match system across backend, Beacon plugin, and frontend with shared config and standardized API responses.
> 
> - **Backend (Convex HTTP)**: Standardizes responses (`{ success, data|error }`); adds batch endpoints `GET /matches/info`, `GET /matches/tokens`, `GET /matches/readiness`; updates `POST /matches/update` to accept `updates[]`; adds `POST /tokens/clear` for reclaiming tokens when players disconnect; switches queries to `matches.getMatches`, `tokens.getTokens`, `tokens.checkReadiness`; imports tokens-per-team and status helpers from `@packages/shared`.
> - **Shared constants**: Introduces `@packages/shared` with `constants/game-config.json` and TS types; Beacon adds `GameConfig.java` (kept in sync with JSON).
> - **Beacon plugin**: Replaces raw `HttpURLConnection` with `ConvexClient`/`HttpClient`/`ConvexResponseParser`; `MatchPollingService` rewritten to batch fetch readiness/tokens/info and start matches; `MatchTelemetryService` batches state updates, queues final telemetry; `MatchManager.endMatch` simplified; `LoginCommand` uses new validate route and tracks `matchId`; `PlayerEventListener` clears tokens on quit via `/tokens/clear`.
> - **Frontend**: Re-exports shared constants in `src/lib/match-constants.ts`; updates practice UI token count and server action to handle standardized response wrapper.
> - Minor: config/deps updates (`@packages/shared` added; lockfile changes).
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 5fd73ba6b23d9e7aa5007ff282d8b289511f992c. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

haoteng10 committed:
Enhance match system and API response handling

- Added support for shared constants and standardized API response formats across the backend and frontend.
- Refactored match-related services to utilize new response structures, improving error handling and consistency.
- Updated match polling and telemetry services to use shared configurations for match statuses.
- Implemented token clearing functionality for players disconnecting before match starts, ensuring tokens can be reused.
- Adjusted various components to reflect changes in match token handling and API interactions.

DIodide committed:
Use JIT package pattern so convex's bundler can resolve shared typescript types

DIodide committed:
Explicitly add turbopack root to blockwarriors-next config. Update shared imports

DIodide commented on apps/blockwarriors-next/tsconfig.json:
is this right?

haoteng10 replied:
I don't think those are necessary. Probably need to run `npm install` whenever we have a new package, even the internal ones.

DIodide commented on packages/shared/package.json:
changed exports to JIT pattern

DIodide commented:
Tested locally, works

haoteng10 committed:
Refactor blockwarriors-next configuration files to simplify path mappings

haoteng10 committed:
Remove jsconfig.json from blockwarriors-next as part of configuration cleanup

haoteng10 committed:
Refactor match services and API interactions for improved consistency

- Updated MatchManager and MatchPollingService to utilize a new ConvexClient for API calls, enhancing error handling and response parsing.
- Replaced ApiResponseParser with ConvexResponseParser across various services to standardize API response handling.
- Enhanced telemetry service to capture and send match state updates more efficiently.
- Implemented batch updates for match statuses and states, allowing multiple updates in a single API call.
- Removed deprecated ApiResponseParser class to streamline the codebase.

haoteng10 commented on apps/blockwarriors-beacon/src/main/java/ai/blockwarriors/beacon/service/MatchManager.java:
```diff
@@ -91,9 +91,9 @@ public void endMatch(String matchId, String winnerPlayerId, UUID deadPlayerId) {
+        // Capture and queue final telemetry NOW while player states are valid
         if (telemetryService != null) {
-            telemetryService.sendFinalMatchState(matchId, winnerPlayerId, deadPlayerId);
+            telemetryService.queueFinalMatchState(matchId, winnerPlayerId);
```
The current implementation is more correct.

haoteng10 committed:
Refactor MatchManager and LoginCommand to utilize ConvexClient for API interactions

- Replaced direct HTTP connection handling with ConvexClient in MatchManager and LoginCommand, improving code readability and maintainability.
- Simplified endMatch method by removing the deadPlayerId parameter, streamlining match termination logic.
- Enhanced error handling and response parsing for login validation, ensuring consistent player experience.
- Updated MatchEventListener and PlayerEventListener to reflect changes in endMatch method signature.
